src/main/java/io/sqreen/powerwaf/ByteBufferSerializer.java

```diff
@@ -265,54 +265,65 @@ ByteBuffer getFirstUsedPWArgsBuffer() {
         }
 
         static class WrittenString {
-            private final long ptr;
-            private final int utf8len;
+            private final Arena arena;
+            private long ptr;
+            private int utf8len;
 
-            WrittenString(long ptr, int utf8len) {
+            WrittenString(Arena arena) {
+                this.arena = arena;
+            }
+
+            public void release() {
+                arena.cachedWS = this;
+            }
+
+            public WrittenString update(long ptr, int utf8len) {
                 this.ptr = ptr;
                 this.utf8len = utf8len;
+                return this;
             }
         }
 
         /**
          * @param s the string to serialize
          * @return the native pointer to the string and its size in bytes,
-         *         or null if the string is too large
+         * or null if the string is too large
          */
         WrittenString writeStringUnlimited(CharSequence s) {
-            ByteBuffer bytes;
-
-            CharBuffer cb = s instanceof CharBuffer ?
-                    ((CharBuffer) s).duplicate() :
-                    CharBuffer.wrap(s);
-
-            try {
-                bytes = getCharsetEncoder().encode(cb);
-            } catch (CharacterCodingException e) {
-                // should not happen
-                throw new UndeclaredThrowableException(e);
+            CharBuffer cb;
+            if (s instanceof CharBuffer) {
+                cb = ((CharBuffer) s).duplicate();
+            } else {
+                cb = currentWrapper = CharSequenceWrapper.wrap(s, currentWrapper);
             }
-            int lenInBytes = bytes.limit();
-            if (lenInBytes == Integer.MAX_VALUE) {
+
+            long tmp = (long) s.length() * MAX_BYTES_PER_CHAR_UTF8 + 1; // 0 terminated
+            if (tmp > Integer.MAX_VALUE) {
                 // overflow ahead
                 return null;
             }
+            int maxBytes = (int) tmp;
 
             StringsSegment segment;
             segment = stringsSegments.get(curStringsSegment);
-            long ptr;
-            while ((ptr = segment.writeNulTerminated(bytes)) == NULLPTR) {
+            WrittenString str = cachedWS;
+            if (str == null) {
+                cachedWS = str = new WrittenString(this);
+            }
+            while ((str = segment.writeNulTerminated(str, getCharsetEncoder(), cb, maxBytes)) == null) {
                 segment = changeStringsSegment(
-                        Math.max(STRINGS_MIN_SEGMENTS_SIZE, lenInBytes + 1 /* NUL */));
+                        Math.max(STRINGS_MIN_SEGMENTS_SIZE, maxBytes));
+                str = cachedWS;
             }
-            return new WrittenString(ptr, lenInBytes);
+            cachedWS = null;
+            return str;
         }
 
-        PWArgsArrayBuffer allocatePWArgsBuffer(int num) {
+        PWArgsArrayBuffer allocatePWArgsBuffer(int num, boolean offsetAddress) {
```

Comment:
I can't tell what the name of the parameter has to do with what it does. It seems to be just bypassing an object pool (and I don't understand why).

Reply:
I agree that the variable name is very bad. I'll try to come up with something better and add comments.

The reason is that the _first_ `ByteBuffer` of the thing that gets serialized, is passed into native code that assumes that it can access the address of the written bytes by calling `GetDirectBufferAddress`, which is not true since the other args are written with an offset from the beginning of the `ByteBuffer`.

Reply:
OK, but is this some insurance policy kind of thing? AFAICT, the first call to `allocatePWArgsBuffer` would regardless return a `PWArgsArrayBuffer` where `start == 0` because the position of the buffer of the first segment will have been reset to 0 by segment.clear() -> buffer.clear(). If `start == 0`, then the address of the bytebuffer will coincide with the start of the data.

Reply:
The tests use the same lease and do multiple calls to `seralizeMore` and then into native code. Those will fit into the same segment and not have buffers that are _get address compatible_.